compiler/rustc_monomorphize/src/collector/abi_check.rs

```diff
@@ -0,0 +1,119 @@
+use rustc_abi::Abi;
+use rustc_middle::ty::{self, Instance, InstanceKind, Ty, TyCtxt};
+use rustc_span::{def_id::DefId, Span, Symbol};
+use rustc_target::abi::call::{FnAbi, PassMode};
+
+use crate::errors::{AbiErrorDisabledVectorTypeCall, AbiErrorDisabledVectorTypeDef};
+
+const SSE_FEATURES: &'static [&'static str] = &["sse", "sse2", "ssse3", "sse3", "sse4.1", "sse4.2"];
+const AVX_FEATURES: &'static [&'static str] = &["avx", "avx2", "f16c", "fma"];
+const AVX512_FEATURES: &'static [&'static str] = &[
+    "avx512f",
+    "avx512bw",
+    "avx512cd",
+    "avx512er",
+    "avx512pf",
+    "avx512vl",
+    "avx512dq",
+    "avx512ifma",
+    "avx512vbmi",
+    "avx512vnni",
+    "avx512bitalg",
+    "avx512vpopcntdq",
+    "avx512bf16",
+    "avx512vbmi2",
+];
+
+fn do_check_abi<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    abi: &FnAbi<'tcx, Ty<'tcx>>,
+    target_feature_def: DefId,
+    emit_err: impl Fn(&'static str),
+) {
+    // This check is a no-op on non-x86, at least for now.
+    if tcx.sess.target.arch != "x86" && tcx.sess.target.arch != "x86_64" {
+        return;
+    }
+    let codegen_attrs = tcx.codegen_fn_attrs(target_feature_def);
+    for arg_abi in abi.args.iter().chain(std::iter::once(&abi.ret)) {
+        let size = arg_abi.layout.size;
+        if matches!(arg_abi.layout.abi, Abi::Vector { .. })
+            && matches!(arg_abi.mode, PassMode::Direct(_))
+        {
+            let features: &[_] = match size.bits() {
+                x if x <= 128 => &[SSE_FEATURES, AVX_FEATURES, AVX512_FEATURES],
+                x if x <= 256 => &[AVX_FEATURES, AVX512_FEATURES],
```

Comment:
A vector of size e.g. 192 should probably also check AVX? I would assume LLVM puts that into an AVX register when it is available, since an SSE register does not suffice.

Reply:
I believe this is what the code is doing - or am I missing something?

Reply:
Ah yes it is, sorry.